### PR TITLE
feat: IFC graph validation

### PR DESCRIPTION
**Pull request by jsantell**
This implements a graph traversal over execution nodes, validating labels throughout according to a policy. This operates over the abstract `PortGraph` trait in the interim as we design the runtime side of things over time.

* Traitify IFC components -- turning `Policy`, `Label`, `Context` etc. into traits allows unit tests to continue testing their domain as we iterate over what the final label/contexts look like without having to update tests across the system, and keeps things flexible. 
* IFC components now operate exclusively on `Label`, rather than `Data<T>` -- `Data<T>` could live in `runtime` in the future possibly
* Itemized PortGraph tests/errors: A consequence of using a trait-based interface with string references is the implementor may define an invalid graph. Future revisions could change this, but in the mean time, checking graph integrity is handled during graph validation.
* Updated Confidentiality/Integrity variants in line with #105

validation algorithm:

```rust
// Seed root_node's outputs from
// data INPUTS
for output in root_node.outputs {
  set_label(output, INPUTS[output])
}

queue.push(root_node)

while queue.len() {
  node = queue.pop()
  if !is_root(node) {
    if !all_inputs_labeled(node) {
      queue.push(node)
      continue
    }
    if all_outputs_labeled(node) {
      continue
    }
    
    // Generate a label for all outputs
    // based off of all inputs
    label = constrain_inputs(node)
    for output in node.outputs {
      set_label(output, label)
    }
  }
  
  for edge in outgoing_edges(node) {
    label = edge.source.label
    // Do not validate cyclic connections
    // back to root
    if !is_root(edge.target) {
      // Get requirements from policy
      // for the label of node's output
      requirements = policy.get(label)
      // Validate that the input node's context
      // satisfies the output port's label based
      // on policy 
      target_ctx = edge.target.context()
      if !requirements.validate(target_ctx) {
        FAIL
      }
    }
    set_label(edge.target.input, label)
  }
  
  // Iterate over outgoing connections again
  // and queue up any connected nodes that
  // are not yet complete. Could probably move
  // the `constrain_inputs` somewhere around here
  // to simplify
  for edge in outgoing_edges(node) {
    if !is_root(edge.target) && !all_outputs_labeled(edge.target) {
      queue.push(edge.target)
    }
  }
}

// All labels on all nodes should be processed at this point.
// If the labels from root_node's inputs differ from output nodes,
// it may be valid, but believe the process should occur once more
// after constraining root_node's inputs -> outputs.
```

### CODE REVIEW

File: rust/common-ifc/src/graph/graph_builder.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    pub fn id(&self) -> &PortGraphId {
+        &self.id
+    }
+
+    pub fn inner(&self) -> &T {
+        &self.inner
+    }
+}
+
+/*
```
Comment (cdata) [inline rust/common-ifc/src/graph/graph_builder.rs:30]: Remove of leave a note

File: rust/common-ifc/src/ifc/validation.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    // Queue empty! At this point, as only the root node may receive cyclic input,
+    // the root node should have it's inputs labeled, and can compare to
+    // the initial input.
+
+    if !are_all_ports_labeled::<G, P::Label>(&label_store, root_node, PortType::Input)? {
+        return Err("Root node has incomplete inputs!".into());
+    }
+
+    debug!("CACHE {:#?}", label_store);
```
Comment (cdata) [inline rust/common-ifc/src/ifc/validation.rs:181]: Clean the prose up a little

**Review by cdata — COMMENTED** [review]
Looking really good!

File: rust/common-ifc/src/graph/graph_builder.rs (hunk 1 above)
Comment (jsantell) [reply rust/common-ifc/src/graph/graph_builder.rs:30]: This entire file shouldn't be here, oops

File: rust/common-ifc/src/context.rs [hunk 3]
```diff
@@ -13,25 +15,41 @@ pub enum ModuleEnvironment {
...
 }
 
-impl Context {
-    /// Ensures the provided [Context] surpasses
-    /// the threshold for all of this context's requirements.
-    pub fn validate(&self, ctx: &Context, input_name: &str) -> Result<()> {
-        if self.env > ctx.env {
-            return Err(CommonIfcError::InvalidEnvironment(input_name.into()));
+impl IfcContext for Context {
+    type Error = String;
```
[the remark this answers is not in the document]
Comment (cdata) [reply rust/common-ifc/src/context.rs:44]: We haven't been using `anyhow` for the same reason, though

**Comment by jsantell** [comment]
superceded by #147

**Pull request closed without merging**